rosidl_typesupport_opensplice_c/resource/msg__type_support_c.cpp.em

```diff
@@ -231,7 +237,14 @@ convert_ros_to_dds_@(__ros_msg_type_prefix)(const void * untyped_ros_message, vo
         return "string not null-terminated";
       }
 
+@[      if isinstance(member.type.value_type, AbstractString)]@
       dds_message->@(member.name)_[i] = DDS::string_dup(str->data);
+@[      else]@
+      std::wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t> cv;
+      dds_message->@(member.name)_[i] = cv.to_bytes(reinterpret_cast<char16_t *>(str->data)).c_str();
```

Comment:
Maybe add a `static_assert` to make sure the types have the same size and alignment requirements?
```c++
static_assert(sizeof(char16_t) == sizeof(std::remove_pointer<decltype(str->data)>::type()) && alignof(char16_t) == alignof(std::remove_pointer<decltype(str->data)>::type()));
```

Alternatively, [this gist claims](https://gist.github.com/shafik/848ae25ee209f698763cffee272a58f8#how-do-we-type-pun-correctly) compilers can optimize out a call to `std::memcpy` so it is the same as `reinterpret_cast`. It seems like a reasonable claim since https://en.cppreference.com/w/cpp/string/byte/memcpy says

> Where strict aliasing prohibits examining the same memory as values of two different types, std::memcpy may be used to convert the values.

Reply:
Good point. I added the asserts in 648d696230bbc4710d0bfb68bf2c6c8b5c74bc4c.